### PR TITLE
refactor(federate-controller): refactor federate controller

### PR DESCRIPTION
**Pull request by limhawjia**

### CODE REVIEW

File: pkg/controllers/federate/controller.go [hunk 1]
```diff
@@ -93,117 +89,147 @@ func (c *FederateController) IsControllerReady() bool {
...
 	dynamicClient dynamicclient.Interface,
-	federatedObjectInformer informers.GenericInformer,
-	sourceObjectInformer informers.GenericInformer,
+	fedClient fedclient.Interface,
+	fedObjectInformer fedcorev1a1informers.FederatedObjectInformer,
+	clusterFedObjectInformer fedcorev1a1informers.ClusterFederatedObjectInformer,
+	informerManager informermanager.InformerManager,
 	metrics stats.Metrics,
 	workerCount int,
 	fedSystemNamespace string,
```
Comment (mrlihanbo) [inline pkg/controllers/federate/controller.go:100]: It seems like `fedSystemNamespace` is unused, maybe we can remove it?

Comment (limhawjia) [reply pkg/controllers/federate/controller.go:102]: I meant to use it to filter out the `kube-admiral-system` objects which we should not federate. Updated the code to include that logic.

File: pkg/util/naming/naming.go [hunk 2]
```diff
@@ -99,7 +99,16 @@ func transformObjectName(objectName string) (string, bool) {
 		}
 	}
 
-	return string(transformedName), transformed
+	// squash any sequence of more than one '.'
+	santizedName := []byte{}
+	for i, ch := range transformedName {
+		if i != 0 && transformedName[i-1] == '.' && transformedName[i] == '.' {
+			continue
```
[the remark this answers is not in the document]
Comment (limhawjia) [reply pkg/util/naming/naming.go:108]: Fixed.

File: pkg/util/cluster/client.go [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+	CertificateAuthorityKey = "certificate-authority-data"
+)
+
+// Service account keys
+const (
+	ServiceAccountTokenKey = "service-account-token-data"
+	ServiceAccountCAKey    = "service-account-ca-data"
+)
+
+func BuildClusterConfig(
```
Comment (gary-lgy) [inline pkg/util/cluster/client.go:45]: Are these copied without modification? What happens to the original file?

File: pkg/util/naming/naming_test.go [hunk 4]
```diff
@@ -33,6 +49,14 @@ func TestGenerateFederatedObjectName(t *testing.T) {
 			},
 			want: "system.foo-roles.rbac.authorization.k8s.io-2728495308",
 		},
+		{
+			name: "generate federated object name with consecutive :",
```
Comment (gary-lgy) [inline pkg/util/naming/naming_test.go:53]: Can we also test consecutive dots?

File: pkg/controllers/federate/util.go [hunk 5]
```diff
@@ -60,115 +72,123 @@ func templateForSourceObject(sourceObj *unstructured.Unstructured, annotations,
...
 
-	if err := unstructured.SetNestedMap(
-		fedObj.Object,
-		templateObject,
-		common.SpecField,
-		common.TemplateField,
-	); err != nil {
-		return nil, err
+	templateObject := templateForSourceObject(sourceObj, templateAnnotations, templateLabels).Object
+	rawTemplate, err := json.Marshal(templateObject)
```
[the remark this answers is not in the document]
Comment (limhawjia) [reply pkg/controllers/federate/util.go:117]: json

File: pkg/controllers/federate/util.go [hunk 6]
```diff
@@ -178,6 +198,19 @@ func updateFederatedObjectForSourceObject(
 		},
 	)
 
+	// Record the observed label and annotation keys in an annotation on the FederatedObject.
+
+	observedAnnotationKeys := generateObservedKeys(sourceObject.GetAnnotations(), federatedAnnotations)
```
[the remark this answers is not in the document]
Comment (limhawjia) [reply pkg/controllers/federate/util.go:203]: Good point, it is probably safer to have it at the start

File: pkg/controllers/federate/controller.go [hunk 7]
```diff
@@ -93,117 +91,160 @@ func (c *FederateController) IsControllerReady() bool {
...
-			}
+			fedObj := obj.(*fedcorev1a1.FederatedObject)
+			return fedObj.Namespace != fedSystemNamespace
+		},
+		Handler: eventhandlers.NewTriggerOnAllChanges(func(o runtime.Object) {
+			fedObj := o.(*fedcorev1a1.FederatedObject)
+			logger := c.logger.WithValues("federated-object", common.NewQualifiedName(fedObj))
 
-			metaObj, err := meta.Accessor(obj)
+			srcMeta, err := meta.GetSourceObjectMeta(fedObj)
```
[the remark this answers is not in the document]
Comment (limhawjia) [reply pkg/controllers/federate/controller.go:158]: these helper functions didn't exist before the rebase. i shall just use `fedObj.Spec.GetTemplateAsUnstructured` instead

File: pkg/util/cluster/client.go (hunk 3 above)
Comment (limhawjia) [reply pkg/util/cluster/client.go:45]: Yes, I should have deleted the original file.

File: pkg/controllers/federate/util.go [hunk 8]
```diff
@@ -37,12 +36,25 @@ import (
...
 
-func templateForSourceObject(sourceObj *unstructured.Unstructured, annotations, labels map[string]string) *unstructured.Unstructured {
+type workerKey struct {
+	name      string
+	namespace string
+	gvk       schema.GroupVersionKind
+}
+
+func (k workerKey) String() string {
+	return fmt.Sprintf("%s/%s", k.namespace, k.name)
```
[the remark this answers is not in the document]
Comment (limhawjia) [reply pkg/controllers/federate/util.go:51]: renamed it to `ObjectKey`

File: pkg/controllers/federate/controller.go [hunk 9]
```diff
@@ -272,6 +273,37 @@ func (c *FederateController) reconcile(ctx context.Context, key workerKey) (stat
 		fedObject = fedObject.DeepCopyGenericFederatedObject()
 	}
 
+	if fedObject != nil {
+		// To account for the very small chance of name collision, we verify the owner reference before proceeding.
+		ownedbySource := false
+		requiresUpdate := -1
```
Comment (gary-lgy) [inline pkg/controllers/federate/controller.go:279]: We don't need to set the OR here. The update logic should already take care of that.

Comment (limhawjia) [reply pkg/controllers/federate/controller.go]: Fixed

**Review by mrlihanbo — APPROVED** [review]

**Pull request merged**